Improve boxed dispatch performance

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33313 Improve boxed dispatch performance**

Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

smessmer committed:
Improve boxed dispatch performance

Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

dzhulgakov commented on c10/util/Bitset.h:
```diff
@@ -0,0 +1,62 @@
+  template <class Func>
+  void for_each_set_bit(Func&& func) const {
+    bitset cur = *this;
+    size_t index = ffsll(cur.bitset_);
```
you can try to use __builtin_ffs - might be faster

smessmer replied:
Why do you think it might be faster? Isn't ffsll internally just calling __builtin_ffsll?

smessmer replied:
I wasn't able to find those calls in the assembly (does that maybe mean they're not inlined???) but the `getDispatchKeyBoxed` function generated exactly the same assembly with `__builtin_ffsll` than with `ffsll`

dzhulgakov replied:
That should be good then - I was posting very low-confidence comments to see whether perf regression comes from there. Given that we found the source of regression - we can disregard it.

dzhulgakov commented on aten/src/ATen/core/dispatch/DispatchKeyExtractor.h:
```diff
@@ -98,15 +99,25 @@ namespace detail {
 
     DispatchKeySet ks;
-    for (const auto& ivalue : torch::jit::last(*stack, num_args_)) {
+    dispatch_arg_indices_reverse_.for_each_set_bit([&] (size_t reverse_arg_index) {
```
did you check the assembly that all these fancy lambdas get inlined? (I still don't trust compilers :) )

smessmer replied:
P126643185, looks inlined to me

dzhulgakov replied:
Yep, looks good as calls to isTensorList appear there nicely.

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]

smessmer committed:
Update on "Improve boxed dispatch performance"


Instead of just remembering the number of arguments and iterating over the stack,
the DispatchKeyExtractor now remembers the exact locations of the dispatch relevant arguments
(i.e. Tensor arguments) and only looks at those.

Benchmarks show a 6% improvement for a TorchScript model calling avgpool1d (which is an op with many non-tensor arguments) in a tight loop.

Differential Revision: [D19748549](https://our.internmc.facebook.com/intern/diff/D19748549/)

[ghstack-poisoned]
